akka-stream/src/main/scala/akka/stream/impl/NewLayout.scala

```diff
@@ -0,0 +1,627 @@
+   * a sequence of steps required to "reconstruct" the graph.
+   *
+   * "Reconstructing" a graph here has a very clear-cut definition: assign a gapless range of integers from
+   * 0..connectionCount to inputs and outputs of modules, so that those that are wired together receive the same
+   * number (and those which are not receive different numbers). This feature can be used to
+   *  - materialize a graph, using the slots as indices to an array of Publishers/Subscribers that need to be wired
+   *    together
+   *  - fuse a graph, using the slots to construct a [[akka.stream.impl.fusing.GraphInterpreter.GraphAssembly]] which
+   *    uses a similar layout
+   *  - create a DOT formatted output for visualization
+   *  - convert the graph to another data structure
+   *
+   * The Traversal is designed to be position independent so that multiple traversals can be composed relatively
+   * simply. This particular feature also avoids issues with multiply imported modules where the identity must
+   * be encoded somehow. The two imports don't need any special treatment as they are at different positions in
+   * the traversal. See [[MaterializeAtomic]] for more details.
+   */
+  sealed trait Traversal {
+
+    /**
+     * Gives the next step of the traversal.
+     */
+    def next: Traversal
+
+    /**
+     * Concatenates two traversals building a new Traversal which traverses both.
+     */
+    def concat(that: Traversal): Traversal = {
+      if (that eq EmptyTraversal) this
+      else Concat(this, that)
+    }
+  }
+
+  /**
+   * A Traversal that consists of two traversals. The linked traversals must be traversed in first, next order.
+   */
+  final case class Concat(first: Traversal, next: Traversal) extends Traversal
+
+  /**
+   * Arriving at this step means that an atomic module needs to be materialized (or any other activity which
+   * assigns something to wired output-input port pairs).
+   *
+   * The traversing party must assign port numbers in the following way:
+   *  - input ports are implicitly assigned to numbers. Every module's input ports are assigned to consecutive numbers
+   *    according to their order in the shape. In other words, the materializer only needs to keep a counter and
+   *    increment it for every visited input port.
+   *  - the assigned number of the first input port for every module should be saved while materializing the module.
+   *    every output port should be assigned to (base + outToSlots(out.id)) where base is the number of the first
+   *    input port of the module (or the last unused input number if it has no input ports) and outToSlots is the
+   *    array provided by the traversal step.
+   *
+   * Since the above two rules always require local only computations (except a counter) this achieves
+   * positional independence of materializations.
+   *
+   * See [[TraversalTestUtils]] class and the `testMaterialize` method for a simple example.
```

Comment:
might be a problem for scaladoc/javadoc to reference test scope class, use backtick